src/Shared/FileUtilities.cs

```diff
@@ -576,6 +576,26 @@ internal static string MaybeAdjustFilePath(string value, string baseDirectory =
             return shouldAdjust ? newValue.ToString() : value;
         }
 
+        /// <summary>
+        /// Gets the path value that is associated with the specified key in a dictionary with <see cref="string"/> values.
+        /// Normalizes the value as a path.
+        /// </summary>
+        /// <param name="dictionary">The dictionary to search.</param>
+        /// <param name="key">The key to locate.</param>
+        /// <param name="value">When this method returns, the value associated with the specified key normalized as a path, if the key is found; otherwise <see langword="null"/>.</param>
+        /// <returns><see langword="true"/> if the dictionary contains an element that has the specified key; otherwise, <see langword="false"/>.</returns>
+        /// <remarks>Use this method to get paths from dictionaries of properties whose default values may contain backslashes.</remarks>
+        internal static bool TryGetPathValue<TKey>(this IReadOnlyDictionary<TKey, string> dictionary, TKey key, out string value)
```

Comment:
I like this for our internal needs. For the public BuildCheck API we're building we should expose a similar operation @JanKrivanek.

Reply:
@rainersigwald ,do we have any ways to catch it earlier , e.g. during evaluation?
We managed to find this spot, but not sure if Jan had a change to try it: https://github.com/dotnet/msbuild/blob/a9efffc4de9609e9ecb45c6c2d41487a4ed70d6f/src/Build/Evaluation/Expander.cs#L1599

Reply:
Unfortunately I don't think there's any 100% reliable way to know during evaluation whether a property is a path or not, unless it already calls `NormalizePath`.

This ties into the longer-term idea of having some more type data in MSBuild so we could say "this property is a partial path" or "this item doesn't represent a file".